derive/src/lib.rs

```diff
@@ -42,14 +41,12 @@ pub fn derive_pod(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
   proc_macro::TokenStream::from(expanded)
 }
 
-/// Derive the `Zeroable` trait for a struct
+/// Derive the `Zeroable` trait for a struct.
 ///
-/// The macro ensures that the struct follows all the the safety requirements
-/// for the `Zeroable` trait.
-///
-/// The following constraints need to be satisfied for the macro to succeed
-///
-/// - All fields ind the struct must to implement `Zeroable`
+/// The macro ensures that the struct follows all the safety requirements for
+/// the `Zeroable` trait. If your struct has generic fields, the generated
+/// `impl` will have a `T: Zeroable` bound for each generic type `T` of your
+/// struct.
```

Comment:
This says it will add a `T: Zeroable` bound, but for a struct like `struct Foo<T: Iterator> { foo: T::Item }` it looks like it will actually generate a `T::Item: Zeroable` bound (which is more correct, but not what derives generally do).

Reply:
This. I personally prefer what this derive does, but it's important that whatever choice is made, it's done being aware that it's not the stdlib's approach (cc @Lokathor).

I personally prefer the approach here, because the stdlib one is what makes:

```rust
#[derive(PartialEq, Eq)]
struct Ptr<T>(*const T);
```

only yield comparable pointers when the pointees are.

Reply:
This is a tough one.
* I'd generally prefer that the *derives* be "extra picky" because this is an unsafe trait after all, and you *can* always just implement manually if you think you know better than the derive's reasoning.
* I do agree that the standard library is a little bit wrong on this one, and I've been bit before by a standard library derive being too conservative (when I wrote `VolAddress<T,R,W>`).

However, I think that since this is becoming an increasingly used crate **we should have the derives work like the stdlib derives**. Don't want people to have any sudden surprises happening.

Reply:
The advantage of the default approach will be that we'll have correctly-spanned error messages even in older rust versions 🙂